### ISSUE TITLE
Fold DispatchStub into c10 dispatcher

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#29844]
Right now we have two levels of dynamic dispatch: c10 dispatcher does a first level of dispatch, and then DispatchStub does a second level of dispatch. For CUDA/HIP dispatch, DispatchStub's extra dispatch is completely unnecessary. However, for CPU dispatch, DispatchStub supports dispatching to AVX/AVX2/default CPU kernels at runtime depending on what instruction set is supported by the processor. We should adjust c10 dispatcher to support this use-case and eliminate DispatchStub entirely.

Related: #29548 (For the rest of this proposal, we assume fallthrough when an operator isn't supported in a dispatch key.)

What is a good encoding for AVX/AVX2/DEFAULT in our dispatcher structure?

1. We should have `DefaultCPUTensorId`, `AVXTensorId` and `AVX2TensorId` to replace `CPUTensorId`. Kernels for code compiled with different AVX settings would register to each of these keys. AVX2 should have highest priority, then AVX, then default CPU.
2. We need some way of ensuring we don't dispatch to AVX2TensorId when the runtime CPU doesn't support AVX2 (and similarly AVX)

There seem to be two ways to go about doing (2).

**Exclude ineligible kernels.** Currently, our computation of the eligible tensor type set is `(tensor_type_set | included_type_set) - excluded_type_set`, where `included_type_set` and `excluded_type_set` are thread local state. Let's suppose every tensor that was previously labeled CPUTensorId, is now labeled DefaultCPUTensorId & AVXTensorId & AVX2TensorId.  If there was a way to add `AVX2TensorId` to `excluded_type_set` for all threads, this would be sufficient to prevent dispatch to it. For example, we could add a `global_excluded_type_set` (at the cost of one more bitwise operation at dispatch time), or we could arrange for `excluded_type_set` to always be initialized with AVX2TensorId when AVX2 is not supported (but TLS variables are, in general, initialized lazily, so this would entail an extra conditional test whenever accessing the TLS variable).

A `global_excluded_type_set` would also help @bwasti do performance experiments with disabling VariableTensorId; however, `variable_is_excluded` would have to be adjusted to check both the TLS and global excluded type set.

**Only put valid CPU capabilities on CPUTensor.** Alternately, instead of labeling every tensor with DefaultCPUTensorId & AVXTensorId & AVX2TensorId, we can only label tensors according to the set of valid capabilities. E.g., on a machine with AVX support only, cpu tensors would only have DefaultCPUTensorId & AVXTensorId. It is a little weird to have to make this change to every tensor, but in this case we wouldn't have to make any modifications to tensor type set computation.

**Comment by zdevito** [comment]
What stops us from just having 1 CPUTensorID, and then decide which kernel to register _at program initialization_. In other words, we just change initialization to be:

```
if (AVX2_exists) {
  register(aten::add, avx2_add);
} else {
  register(aten::add, normal_add);
}
```

This wasn't possible with vtables, but now that we control what is in the tables, can't we just register differently to resolve this?  This also resolves what to do for kernels that don't have implementations in AVX2 -- simply don't register them.

**Comment by ezyang** [comment]
This seems like it should work :>

**Comment by smessmer** [comment]
I agree with @zdevito. AVX doesn't seem like a property of a tensor, you don't want to dispatch some tensors to AVX and some to non-AVX like you might do with CPU and CUDA.

**Comment by dzhulgakov** [comment]
Agree for cuda/hip dispatch.

For AVX - I'm not fully sold it should happen on c10::Dispatcher level. There's usually big enough shared part of CPU code (all the checks, allocations, etc) and only the inner compute kernel is avx/avx2 dependent. So having a separate lightweight mechanism for it (DispatchStub) feels more convenient. FWIW, there's a similar mechanism in caffe2/perfkernels - we should probably keep only one of them

**Comment by ezyang** [comment]
> only the inner compute kernel is avx/avx2 dependent. So having a separate lightweight mechanism for it (DispatchStub) feels more convenient

I mean, I think this is a problem we should generally solve: shared code that should be reused between different implementations of the kernel (this applies across cpu and cuda).  Also, there isn't all that much that's shared in practice:

```
  static Tensor threshold_out(
      optional<Tensor> opt_result,
      const Tensor& self,
      Scalar threshold,
      Scalar value,
      const Tensor& other) {
    Tensor result = opt_result.value_or(Tensor());
    auto iter = TensorIterator::binary_op(result, self, other);
    threshold_stub(iter.device_type(), iter, threshold, value);
    return iter.output();
  }
```

**Comment by ezyang** [comment]
Some updates to this issue:
1. I've come around to @dzhulgakov's point that there is shared CPU code that should be deduplicated. So my idea now is that we should still have a two step dynamic dispatch; the only difference is that we should use c10 dispatcher to do both levels of dispatch. There isn't really a technical reason to do this (in some sense, DispatchStub is superior at what it does), it is mostly to make it easier for people to understand the code and not have to scratch their heads about there being two dispatch mechanisms.
2. We're going to do the moral equivalent of Zach's if statement, but we're going to implement it with priorities so that we can keep static initializing the kernels separately

**Comment by ezyang** [comment]
But then, going one step further, maybe there is no benefit in folding DispatchStub into c10 dispatcher, and it should stay so we can keep c10 dispatcher oblivious of vectorization concerns.

**Comment by ezyang** [comment]
Now that we have structured kernels https://github.com/pytorch/pytorch/issues/55070 we don't need two levels as we can duplicate the prologue automatically! So it is now entirely a binary size question.

**Comment by jgong5** [comment]
> Now that we have structured kernels #55070 we don't need two levels as we can duplicate the prologue automatically! So it is now entirely a binary size question.

@ezyang Do you have a working example how kernels implemented with AVX2 and AVX512 can be enabled and dispatched via the new structured kernel approach? Thanks.

**Comment by ezyang** [comment]
This issue describes a hypothetical feature that isn't implemented. Today vectorized kernels still get funneled through DispatchStub even if they are structured
